query/src/datasources/table_func/numbers_stream.rs

```diff
@@ -108,6 +118,11 @@ impl NumbersStream {
 
             let series = DFUInt64Array::new_from_aligned_vec(av).into_series();
             let block = DataBlock::create_by_array(self.schema.clone(), vec![series]);
+            if !self.sort_columns_descriptions.is_empty() {
```

Comment:
This did not help to block generate.

select number from numbers(1000) order by number desc limit 3;

`SortPartialTransform` already did the limit improvement for us.

We should push down the sort/limit into `try_get_one_block`.

Reply:
"We should push down the sort/limit into try_get_one_block."
-- This change is actually inside of function try_get_one_block. Or are you suggesting change the BlockRange ? 

Also @sundy-li , I have some questions related to this task:
1. Do we have plan to support query like "order by Func(number)". The push down to BlockRange can only apply for "order by number" right ? If we have plan, the execution of "Func(number)" can be pushed down to remote machine, right ? 
2. Another question is, is the "SortPartialTransform" executed on local machine ? I am thinking, the read_table can be executed in remote machine, thus it can decrease the amount of data to transfer right ?

Reply:
> This did not help to block generate.
> 
> select number from numbers(1000) order by number desc limit 3;
> 
> `SortPartialTransform` already did the limit improvement for us.
> 
> We should push down the sort/limit into `try_get_one_block`.

It is actually inside of function try_get_one_block. I think at least one of the benefit is, it decreases the data block size to transmit, right ? Every thread just send  limit number of rows, right ?

Reply:
> I think at least one of the benefit is, it decreases the data block size to transmit, right
Yes, currently it benefits the data transfer to other nodes. It saves io mostly and it's better than previous.

But if you cut to limit the block, it can save io & CPU both, it's even better.

Reply:
The number generation in try_get_block is in sequence, so it's easy to apply sort & limit pushdown optimization.

Reply:
sure, will address

Reply:
For other functions like:  `order by number + 1`, `order by (number * 3)`, this is related to #2343.

In this feature, we just make it simple using column name match.

Reply:
Thanks! Yes, the monotonic check is something I thought about. And I thought you the logical would be implemented in the DataBlock sort, That is why I chose use DataBlock sort, instead of jut changing DataRange in the PR.
Thanks for clarifying, will address accordingly.

BTW, @sundy-li @BohuTANG , someone with investing background sent me a message, asking about the creator of this project. I forwarded the message to you in the Slack, could you take a look ?

Reply:
Sorry for later, sure :)
Thank you